[REVIEW] Parquet reader - add support for multiple sources

vuule opened the pull request:
Fixes #3841 

- Adds ctors to `source_info` that take vectors, so that the info object describes multiple sources.
- Parquet reader can now read from multiple sources into a single table.
- Sources need to have matching schemas for now, we can relax this later to only match the selected columns.
- Modify the Python implementation to also allow passing a list of sources and a list of row group lists. Users can still pass a single source/row group list.
- Removed `row_group` and `row_group_size` based on @kkraus14 suggestion.
- refactor the Parquet `metadata` struct and add `aggregate_metadata` to represent the metadata of the entire source set.
- Added tests for the new capabilities, and negative tests for mismatching input/parameters.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

jakirkham commented:
cc @rjzamora (in case this is of interest 😉)

rjzamora commented on python/cudf/cudf/io/parquet.py:
```diff
@@ -164,8 +164,7 @@ def read_parquet(
     columns=None,
-    row_group=None,
-    row_group_count=None,
+    row_group_list=None,
```
I'm a big fan of this change :)

kkraus14 replied:
Can we rename this to `row_groups` instead? Presumably we can still pass a scalar integer here?

vuule replied:
done

rjzamora commented on python/cudf/cudf/tests/test_parquet.py:
```diff
@@ -281,13 +282,16 @@ def num_row_groups(rows, group_size):
 
-    gdf = [cudf.read_parquet(fname, row_group=i) for i in range(row_groups)]
+    gdf = [
+        cudf.read_parquet([fname, fname], row_group_list=[[i], []])
```
If it's not too much of a pain, it would be nice to test the new `row_group_list` mechanism for a single file, and for non-contiguous lists of row-group indices.  For example, in the near future, `dask.dataframe`/`dask_cudf` will likely want to pass in something like `row_group_list=[0,1,3,8,9]` for a single file.

vuule replied:
added a test that alternates between the two sources. Modified the basic row group list test back to a single source, it did not increase code coverage with the newly added cases.

rjzamora commented on python/cudf/cudf/tests/test_parquet.py:
```diff
@@ -298,6 +302,30 @@ def test_parquet_read_row_group(tmpdir, pdf, row_group_size):
+    if "col_category" in gdf.columns:
+        gdf = gdf.drop("col_category")
+
+    assert_eq(pdf.reset_index(drop=True), gdf, check_categorical=False)
```
Do you need to drop the categorical column **and** use `check_categorical=False` here?

vuule replied:
Apparently not, thanks! Modified all instances where this was done.

rjzamora commented on python/dask_cudf/dask_cudf/io/opt_parquet.py:
```diff
@@ -124,18 +124,12 @@ def _read_partition(part, index, columns, strings_to_cats):
     if isinstance(index, list):
         columns += index
 
```
Note that the primary `read_parquet` code path for `dask_cudf` is in a separate file (you cant think of the code in `opt_parquet.py` as a temporary/optimized code path to avoid some slow upstream behavior when the user passes in `row_groups_per_part`).  The official path is still passing `row_group` in a few places [like here](https://github.com/rapidsai/cudf/blob/1a52851a0b447b26318602d6f4de8e8b4febdfe0/python/dask_cudf/dask_cudf/io/parquet.py#L55) (which is likely causing the CI failures).

vuule replied:
found it, thanks!

rjzamora commented on python/dask_cudf/dask_cudf/io/opt_parquet.py:
```diff
@@ -124,18 +124,12 @@ def _read_partition(part, index, columns, strings_to_cats):
-        )
+    df = cudf.io.read_parquet(
+        path,
+        row_group_list=row_groups,
```
When there is no `_metadata` file available, `row_groups` can theoretically be passed as `None` here.  Just want to confirm that `row_group_list` can handle a `None` input?

vuule replied:
Yup, that's fine. None is the default value.

kkraus14 commented on python/cudf/cudf/_lib/io/utils.pyx:
```diff
@@ -15,31 +17,46 @@ import os
-    cdef const unsigned char[::1] buf
+    if not len(src):
+        raise ValueError("Need to pass at least one source")
+    print(type(src[0]))
```
Debug print

vuule replied:
removed, thanks!

kkraus14 commented on python/cudf/cudf/io/parquet.py:
```diff
@@ -175,29 +174,47 @@ def read_parquet(
+    # a list of row groups per source should be passed. If a single list is
+    # passed, wrap in into a list to make the list of lists that is expected
+    # for multiple sources
+    if row_group_list and not isinstance(row_group_list[0], list):
+        row_group_list = [row_group_list]
```
If `row_group_list` is a scalar say `1` here the `[0]` index will fail here

vuule replied:
yeah, the implementation expects a list, not a scalar. I'll add this case. 
With this, we'll support everything from a scalar to a list of lists :D

vuule replied:
done

kkraus14 commented on python/cudf/cudf/io/parquet.py:
```diff
@@ -175,29 +174,47 @@ def read_parquet(
-        raise ValueError("URL content-encoding decompression is not supported")
+    # Multiple sources are passed as a list. If a single source is passed,
+    # wrap it in a list for unified processing downstream.
+    if not isinstance(filepath_or_buffer, list):
+        filepath_or_buffer = [filepath_or_buffer]
```
This is confusing if someone passes say an `tuple` of filenames, because we'll get a list with a single `tuple` in it instead of a list of filenames. I believe we have a utility function that we can use for `is_list_like`: https://github.com/rapidsai/cudf/blob/e3624d0679cb2c383123fca14286049a65708f2a/python/cudf/cudf/utils/dtypes.py#L211-L226

vuule replied:
thanks for the pointer, useful for both sources and row groups. Fixed the code to sue this.

kkraus14 commented on python/cudf/cudf/io/parquet.py:
```diff
@@ -175,29 +174,47 @@ def read_parquet(
-            filepath_or_buffer, columns=columns, *args, **kwargs
-        )
-        return cudf.DataFrame.from_arrow(pa_table)
+        pa_tables = [
+            cudf.DataFrame.from_arrow(
+                pq.read_pandas(fob, columns=columns, *args, **kwargs)
+            )
+            for fob in filepaths_or_buffers
+        ]
+        return cudf.concat(pa_tables)
```
Does arrow not support passing a vector of files down to it?

vuule replied:
Looks like it does not (at least I could not pass the list directly).
I found [a comment in an Arrow issue](https://issues.apache.org/jira/browse/ARROW-3424?focusedCommentId=17082957&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17082957) that may suggest that there is a way to do this, really not sure.

kkraus14 replied:
Looks like we'd need to do `pq.ParquetDataset(filepaths_or_buffers).read()`: https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetDataset.html

kkraus14 commented on python/cudf/cudf/tests/test_parquet.py:
```diff
@@ -281,21 +282,54 @@ def num_row_groups(rows, group_size):
-    gdf = [cudf.read_parquet(fname, row_group=i) for i in range(row_groups)]
-    gdf = cudf.concat(gdf).reset_index(drop=True)
+    gdf = [
+        cudf.read_parquet(fname, row_group_list=[i]) for i in range(row_groups)
```
Can we keep `i` as a scalar for the row group?

vuule replied:
done

vuule commented:
> Can we rename this to row_groups instead? Presumably we can still pass a scalar integer here?

Somehow I cannot reply or resolve this. Renamed, changed the Python layer to accept scalars.

vuule commented on python/cudf/cudf/_lib/io/utils.pyx:
```diff
@@ -15,31 +17,46 @@ import os
 # Converts the Python source input to libcudf++ IO source_info
 # with the appropriate type and source values
 cdef source_info make_source_info(src) except*:
```
done

karthikeyann commented on cpp/include/cudf/io/types.hpp:
```diff
@@ -137,37 +137,64 @@ struct table_with_metadata {
   {
   }
-};
+};  // namespace io
```
Probably a code formatting error. This is not closing of namespace io.

karthikeyann commented on cpp/src/io/avro/reader_impl.cu:
```diff
@@ -464,19 +464,20 @@ table_with_metadata reader::impl::read(int skip_rows, int num_rows, cudaStream_t
                reader_options const &options,
                rmm::mr::device_memory_resource *mr)
-  : _impl(std::make_unique<impl>(datasource::create(filepath), options, mr))
+  : _impl(std::make_unique<impl>(datasource::create(filepaths[0]), options, mr))
```
similar `CUDF_EXPECTS` should be present. 
`CUDF_EXPECTS(sources.size() == 1, "Only a single source is currently supported.");`

karthikeyann commented on cpp/src/io/csv/reader_impl.cu:
```diff
@@ -758,21 +758,22 @@ reader::impl::impl(std::unique_ptr<datasource> source,
                reader_options const &options,
                rmm::mr::device_memory_resource *mr)
-  : _impl(std::make_unique<impl>(nullptr, filepath, options, mr))
+  : _impl(std::make_unique<impl>(nullptr, filepaths[0], options, mr))
```
similar CUDF_EXPECTS should be present.

vuule replied:
Absolutely. Missed these. Done.

karthikeyann commented on cpp/src/io/functions.cpp:
```diff
@@ -28,24 +28,25 @@ namespace cudf {
+    datasources = cudf::io::datasource::create(src_info.files);
+  } else if (src_info.type == io_type::USER_IMPLEMENTED) {
+    datasources = cudf::io::datasource::create(src_info.user_sources);
+  } else
```
braces for `else` too for uniformity.

vuule replied:
done

karthikeyann commented on cpp/src/io/json/reader_impl.cu:
```diff
@@ -536,21 +536,22 @@ table_with_metadata reader::impl::read(size_t range_offset, size_t range_size, c
                reader_options const &options,
                rmm::mr::device_memory_resource *mr)
-  : _impl(std::make_unique<impl>(nullptr, filepath, options, mr))
+  : _impl(std::make_unique<impl>(nullptr, filepaths[0], options, mr))
```
similar `CUDF_EXPECTS` as other method.

karthikeyann commented on cpp/src/io/orc/reader_impl.cu:
```diff
@@ -833,19 +833,20 @@ table_with_metadata reader::impl::read(size_type skip_rows,
                reader_options const &options,
                rmm::mr::device_memory_resource *mr)
-  : _impl(std::make_unique<impl>(datasource::create(filepath), options, mr))
+  : _impl(std::make_unique<impl>(datasource::create(filepaths[0]), options, mr))
```
similar `CUDF_EXPECTS` as  other method.

karthikeyann commented on cpp/src/io/parquet/reader_impl.hpp:
```diff
@@ -57,7 +57,7 @@ class reader::impl {
    * @param mr Device memory resource to use for device memory allocation
    */
-  explicit impl(std::unique_ptr<datasource> source,
+  explicit impl(std::vector<std::unique_ptr<datasource>> &&sources,
```
documentation update for all argument name changes and meaning changes.

karthikeyann commented on cpp/include/cudf/io/readers.hpp:
```diff
@@ -78,7 +78,7 @@ class reader {
    * @param mr Device memory resource to use for device memory allocation
    */
-  explicit reader(std::string filepath,
+  explicit reader(std::vector<std::string> const &filepaths,
```
doxygen documentation updates.

karthikeyann commented on cpp/include/cudf/io/types.hpp:
```diff
@@ -137,37 +137,64 @@ struct table_with_metadata {
 };
 
+/**
+ * @brief TODO
```
docs here.

ayushdg commented on python/cudf/cudf/io/parquet.py:
```diff
@@ -164,8 +165,7 @@ def read_parquet(
     engine="cudf",
     columns=None,
-    row_group=None,
-    row_group_count=None,
+    row_groups=None,
```
Doctoring needs an update to reflect this change

ayushdg commented on python/cudf/cudf/io/parquet.py:
```diff
@@ -175,29 +175,47 @@ def read_parquet(
+            row_groups = [row_groups]
+
+    filepaths_or_buffers = []
+    for source in filepath_or_buffer:
+        tmp_source, compression = ioutils.get_filepath_or_buffer(
+            source, None, **kwargs
+        )
```
This can be handled in a followup pr: The current logic doesn't work when the user passes a filepath with a wildcard like `path/*.parquet` since we call the `ioutils.get_filepath_or_buffer` method on each string and the logic implemented [here](https://github.com/rapidsai/cudf/blob/branch-0.15/python/cudf/cudf/utils/ioutils.py#L934-L940) guards against passing multiple files (since it was unsupported at the time). This utility can potentially be updated to return a list of files and readers which support this use it while others warn/raise errors.

devavret commented on cpp/src/io/parquet/reader_impl.cu:
```diff
@@ -342,14 +442,16 @@ struct metadata : public FileMetaData {
-                                      size_t end_chunk,
-                                      const std::vector<size_t> &column_chunk_offsets,
-                                      cudaStream_t stream)
+};  // namespace parquet
```
Looks like the close of aggregate_metadata

karthikeyann commented on cpp/include/cudf/io/datasource.hpp:
```diff
@@ -71,10 +72,9 @@ class datasource {
    *
-   * @param[in] data Pointer to the input data host buffer
-   * @param[in] size Size of the buffer in bytes
+   * @param[in] buffer Host buffer object
```
factory `@return` documentation.

karthikeyann commented on cpp/include/cudf/io/datasource.hpp:
```diff
@@ -91,6 +91,22 @@ class datasource {
+  /**
+   * @brief Creates a vector of datasources, one per element in the input vector.
+   *
+   * @param[in] args vector of parameters
```
`@return` documentation of factory method.

kkraus14 commented:
I believe @vuule is on vacation this week where if someone wants to push the last changes to this branch to move this across the finish line it would be appreciated. @karthikeyann @devavret

karthikeyann commented:
rerun tests

jakirkham commented:
Thanks @vuule! 😄
